File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        # inputs: mxn matrix
        # ouptut: an array returning all elements in the matrix in spiral order
        output = []
        top, bottom = 0, len(matrix) - 1
        left, right = 0, len(matrix[0]) - 1

        while top <= bottom and left <= right:
            for i in range(left, right + 1):
                output.append(matrix[top][i])
            top += 1

            for i in range(top, bottom + 1):
                output.append(matrix[i][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    output.append(matrix[bottom][i])
                bottom -= 1
            if left <= right:
                for i in range(bottom, top - 1, -1):
                    output.append(matrix[i][left])
                left += 1

        return output

File: 54-spiral-matrix/spiral-matrix.rb
# @param {Integer[][]} matrix
# @return {Integer[]}
def spiral_order(matrix)
    left = 0
    right = matrix[0].size - 1
    top = 0
    bottom = matrix.size - 1
    output = []

    while left <= right && top <= bottom
        (left..right).each do |i|
            output << matrix[top][i]
        end
        top += 1

        (top..bottom).each do |i|
            output << matrix[i][right]
        end
        right -= 1

        if top <= bottom
            (left..right).reverse_each do |i|
                output << matrix[bottom][i]
            end
            bottom -= 1
        end

        if left <= right
            (top..bottom).reverse_each do |i|
                output << matrix[i][left]
            end
            left += 1
        end
    end

    output
end